libexec/klab-prove

```diff
@@ -105,7 +105,7 @@ timeout "$TIMEOUT" "$KLAB_EVMS_PATH/deps/k/k-distribution/target/release/k/bin/k
   --output json \
   --boundary-cells k,pc \
   --smt_prelude "$KLAB_OUT/prelude.smt2" \
-  --z3-tactic "(or-else (using-params smt :random-seed 3 :timeout 1000) (using-params smt :random-seed 2 :timeout 2000) (using-params smt :random-seed 1))" \
```

Comment:
I had understood from conversations with @mhhf the purpose of this line was to try the z3 solver multiple times with different timeouts. If we remove the timeouts does this line still make sense? Am I misunderstanding it's purpose?

Reply:
my understanding was that z3 would still be invoked with different tactics, but now they would all just run with the default timeout... I'm not alltogether certain about how z3 deals with tactics though... Maybe @leonardoalt or @daejunpark can shed some light on whats going on here

Reply:
To me it sounds like it would still be invoked multiple times, just with different seeds. Does z3 have a default timeout != infinity though?

Reply:
Well there is still the `--z3-impl-timeout 300` above, so K will still kill Z3 if it hangs for longer than 300 microseconds.

Reply:
But I'm not confident in what this means, does it mean it starts 3 instances of the Z3 solver in parallel, or tries them in sequence? If in sequence, it probably will only ever run the first random seed, in which case we might as well not bother with a `--z3-tactic` altogether.

Reply:
I remember we added this in the summer of 2018 when we discovered z3 peformance on certain queries (arithmetic-related iirc) was dependent on the seed: for some seeds it timed out and for some it returned instantly. Thus we added this line to make it try a couple of different seeds with increasing timeouts. 

This was many z3s and evm-semantics and Ks ago, we should see if perhaps this non-determinism is gone now...

Reply:
I'm not sure about arithmetics, but for the stuff that I usually solve (Horn) this seed dependent behavior is definitely there.

Reply:
I agree, especially with an update to Z3, I think it would be good to check if performance is good enough without this tactic.

Reply:
> Well there is still the `--z3-impl-timeout 300` above, so K will still kill Z3 if it hangs for longer than 300 microseconds.

do you mean millieconds or are we really on the micro level here?

Reply:
@MrChico It's [milliseconds](https://github.com/kframework/k/blob/91a69ed56a92897677d16bbdb2065841a4446125/kernel/src/main/java/org/kframework/utils/options/SMTOptions.java#L51)

Reply:
I found the use of this tactic in klab when I helped maker folks to debug their proofs. At that time, I was also curious how the tactic timeout can work in the presence of the K timeout. Both timeouts are in milliseconds, thus if the first instance of the z3 solver does not return immediately, there seems to be no chance of getting to the second instance. But, the more critical problem at that time was that z3 hangs forever with this tactic for some proofs, so we ended up dropping this tactic at least for such proofs. Other than that, I haven't had much experience of using tactics.